### PR TITLE
[ENH] Add support of training to `TimeMoE`

### PR DESCRIPTION
**Pull request by Faakhir30**
#### Reference Issues/PRs

Refs [#10536 comment](https://github.com/sktime/sktime/issues/10536#issuecomment-4948456543)
Started with fixing `config`, but changed direction to add support of training.

#### What does this implement/fix? Explain your changes.

- added more vendored code for pretraining support
- set `"capability:pretrain": True`
- added `_pretrain` method for fine-tuning / training from scratch
- zero-shot forecasting via ``fit`` + ``predict``
- fine-tuning of a pretrained checkpoint via ``pretrain``
- training from scratch via ``pretrain`` with ``model_path=None``
- add examples to docstring
- other minor fixes like correcting param name in `get_test_params`

This fix passes full config passed by user to the model.

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://github.com/all-contributors/all-contributors/blob/master/docs/emoji-key.md)
- [x] Optionally, for added estimators: I've added myself and possibly to the `maintainers` tag - do this if you want to become the owner or maintainer of an estimator you added.
  See here for further details on the [algorithm maintainer role](https://www.sktime.net/en/latest/get_involved/governance.html#algorithm-maintainers).
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

### CODE REVIEW

**Review by geetu040 — CHANGES_REQUESTED** [review]
I would suggest to do the opposite - instead of forwarding full config, don't forward it at all. Which is exactly what's happening in code but misleading in docs. Therefore I would update only the docs to inform the user that other values in config are not used.

I prefer this because, the model doesn't support training so providing a different config is meaningless.

Regarding the parameters, we can discuss if we want to deprecate them or keep just in-case we have this in future.

For now I would say, let's not supply the config until there is training support added in parallel.

Btw was there any discussion around this? or any relevant issue except for https://github.com/sktime/sktime/issues/10536 that I am missing? Because is a slightly different problem and needs to be discussed in a separate issue.

**Comment by Faakhir30** [comment]
> Btw was there any discussion around this? or any relevant issue except for https://github.com/sktime/sktime/issues/10536 that I am missing? Because is a slightly different problem and needs to be discussed in a separate issue.

Nope, opened PR seperatly for this one to have discusion ;)

> the model doesn't support training

hmm, I see that docstring mentions its zero-shot, but I'm suspecting that is not the case

- see section 4.2 in [paper](https://arxiv.org/pdf/2409.16040)
- also, the the codebase suggests it is trainable both from scratch and on smaller dataset on top of pretrained one for few-shot, see readme of [repo](https://github.com/Time-MoE/Time-MoE) 
- also see implementation of `runner.train_model` called in `main.py`

I'll give it a shot for support of training, I think #7869 only considered this as zero-shot forecaster by mistake.

**Comment by geetu040** [comment]
Yeah agreed, I think `transformers.Trainer` should be compatible with this since `TimeMoeForPrediction.forward` can take `labels` as argument and compute the loss internally.

I think `_fit` should remain as it is and we only need to implement `_pretrain`.
Please use `TimesFM2Forecaster` as reference, it would be a relevant example.

**Comment by Faakhir30** [comment]
> I think transformers.Trainer should be compatible with this since TimeMoeForPrediction.forward can take labels as argument and compute the loss internally.

Yeah, Time-MoE package already provides `TimeMoeTrainer(transformers.Trainer)` and dataset class, vendored those trainer+dataset files and used those, thanks

File: sktime/forecasting/timemoe.py [hunk 1]
```diff
@@ -70,6 +77,9 @@ class TimeMoEForecaster(_GlobalForecastingDeprecationMixin, BaseForecaster):
         - tie_word_embeddings: bool, default=False
             Whether to tie word embeddings in the TimeMOE model.
 
+        Architecture keys are used when ``model_path=None`` (from-scratch) and
+        ignored otherwise.
+
```
Comment (geetu040) [inline sktime/forecasting/timemoe.py:87]: "Architecture keys" - this is unclear

File: sktime/forecasting/timemoe.py [hunk 2]
```diff
@@ -86,6 +96,22 @@ class TimeMoEForecaster(_GlobalForecastingDeprecationMixin, BaseForecaster):
...
+        Sliding-window length for ``pretrain``. Defaults to ``1024`` when ``None``.
+        For small datasets, use a shorter length with ``stride=1``.
+
+    stride : int, optional (default=None)
+        Sliding-window stride for ``pretrain``. Defaults to ``context_length``.
+
+    training_args : dict, optional (default=None)
+        Keyword arguments used for training.
+        Supports all arguments by ``transformers.TrainingArguments`` availble at
+        https://huggingface.co/docs/transformers/en/main_classes/trainer#transformers.TrainingArguments.
```
Comment (geetu040) [inline sktime/forecasting/timemoe.py]: better to keep this in the section "References" and then use the number here

File: sktime/forecasting/timemoe.py [hunk 3]
```diff
@@ -181,6 +253,44 @@ def __post_init__(self):
...
+            TimeMoETrainingArguments,
+        )
+        from sktime.libs.timemoe.time_moe_window_dataset import TimeMoEWindowDataset
+        from sktime.libs.timemoe.ts_dataset import SeriesListDataset
+
+        self.model_ = self._load_model()
+
+        context_length = (
+            self.context_length if self.context_length is not None else 1024
+        )
```
Comment (geetu040) [inline sktime/forecasting/timemoe.py]: why not set 1024 as the default value in the parameter?

File: sktime/forecasting/timemoe.py [hunk 4]
```diff
@@ -266,7 +381,7 @@ def _get_default_config(self):
             "apply_aux_loss": True,
             "router_aux_loss_factor": 0.02,
             "tie_word_embeddings": False,
-            "torch_dtype": torch.bfloat16,
+            "torch_dtype": torch.float32,
             "device_map": "cpu",
```
Comment (geetu040) [inline sktime/forecasting/timemoe.py]: `"torch_dtype"` and `"device_map"` should be provided by the user directly and kept away from the config

**Review by geetu040 — CHANGES_REQUESTED** [review]
Nice, implementation is mostly good, I've left a few comments for improvements.

File: sktime/forecasting/timemoe.py (hunk 4 above)
Comment (Faakhir30) [reply sktime/forecasting/timemoe.py]: Done.
I see, these are used for pretrained ones too (zero-shot), and `config` should be used for from-scratch training configuration, right?

Comment (geetu040) [reply sktime/forecasting/timemoe.py]: > I see, these are used for pretrained loaded from a checkpoint too

yes, these should have been out of config from the start, but better to separate them now

> `config` should be used for from-scratch training configuration, right?

not necessarily, when model_path is provided, we should load the config from that path, update it with user-provided config and then load the model. this way only affected weights will be initialized randomly.

File: sktime/forecasting/timemoe.py [hunk 5]
```diff
@@ -70,6 +77,20 @@ class TimeMoEForecaster(BaseForecaster):
         - tie_word_embeddings: bool, default=False
             Whether to tie word embeddings in the TimeMOE model.
 
+        These keys initialize the model architecture when ``model_path=None``
+        (from-scratch). When loading a pretrained checkpoint they are ignored.
+
+        Do not pass ``torch_dtype`` or ``device_map`` here; use the dedicated
+        constructor parameters instead.
```
Comment (geetu040) [inline sktime/forecasting/timemoe.py]: not really needed, should be obvious

File: sktime/forecasting/timemoe.py [hunk 6]
```diff
@@ -70,6 +77,20 @@ class TimeMoEForecaster(BaseForecaster):
...
+
+    device_map : str, dict, int, or torch.device, default="cpu"
+        Device placement following the ``transformers`` ``device_map`` naming
+        convention, for example ``"cpu"``, ``"cuda"``, or ``"auto"``.
+
+    torch_dtype : torch.dtype, optional (default=torch.float32)
+        Torch dtype used when loading the model and preparing prediction
+        inputs. Defaults to ``torch.float32``.
```
Comment (geetu040) [inline sktime/forecasting/timemoe.py]: let's call these `device` and `dtype` to generalize with other foundation models

File: sktime/forecasting/timemoe.py [hunk 7]
```diff
@@ -133,17 +211,27 @@ class TimeMoEForecaster(BaseForecaster):
...
         config: dict = None,
+        device_map="cpu",
+        torch_dtype=None,
         seed: int = None,
         use_source_package: bool = False,
         ignore_deps: bool = False,
+        context_length: int = 1024,
+        stride: int = None,
+        training_args: dict = None,
     ):
```
Comment (geetu040) [inline sktime/forecasting/timemoe.py:233]: new parameters should be added in the last, this breaks backward compatibility

File: sktime/forecasting/timemoe.py [hunk 8]
```diff
@@ -174,12 +262,53 @@ def __post_init__(self):
...
         * any soft dependency imports in the constructor
         """
+        import torch
+
         self._seed = np.random.randint(0, 2**31) if self.seed is None else self.seed
 
+        self._torch_dtype = (
+            torch.float32 if self.torch_dtype is None else self.torch_dtype
+        )
+
```
Comment (geetu040) [inline sktime/forecasting/timemoe.py]: can we avoid this? `torch_dtype=None` when passed during `from_pretrained` should load the model in it's native dtype

File: sktime/forecasting/timemoe.py [hunk 9]
```diff
@@ -381,51 +509,96 @@ def _predict(self, fh, X=None):
...
-        """Load the model from checkpoint."""
+        """Load from checkpoint, or initialize from config when path is absent."""
+        if self.model is not None:
+            return self.model
+
+        if self.timemoe_kwargs.get("pretrained_model_name_or_path") is None:
+            self.model = self._load_from_config()
+        else:
+            self.model = self._load_pretrained()
+        return self.model
```
Comment (geetu040) [inline sktime/forecasting/timemoe.py:630]: when loading from pretrained, config is completely ignored.

hear out this use-case, when somebody wants to use the model from the checkpoint with only slight changes in the config like the head, in current scenario that would not be possible.

we have to choose between 2 options:
- keep it this way and inform through the docstring that config will be ignored when model path is provided
- update the code to use the config and document this behaviour that custom config with model path causes some weights to be reinitialized and model need to be pre-trained.

what do you think about this?

**Review by geetu040 — CHANGES_REQUESTED** [review]

File: sktime/forecasting/timemoe.py (hunk 9 above)
Comment (Faakhir30) [reply sktime/forecasting/timemoe.py:630]: I do agree with the usecase, we should allow custom configs with pretrained models.

Also in pre-trained path, 
- Option 1: I think we should not consider config from `_get_default_config`. It would be better to build using `ConfigClass.from_pretrained(path)` +  (user passed `config` only, no defaults) to respect pretrained model's config.
- Another option could be to proritize in following order: `_get_default_config` follwed by `ConfigClass.from_pretrained(path)` followed by user passed `config`. (slightly harder to debug, so implemented option 1 for now)

File: sktime/libs/timemoe/time_moe_window_dataset.py [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+        self.window_size_plus_one = self.window_size + 1
+        self.stride = stride if stride else self.window_size
+
+        num_seqs = len(self.dataset)
+        iterator = range(num_seqs)
+        try:
+            from tqdm import tqdm
+
+            iterator = tqdm(iterator, total=num_seqs)
+        except ImportError:
```
[the remark this answers is not in the document]
Comment (Faakhir30) [reply sktime/libs/timemoe/time_moe_window_dataset.py:39]: would prefer to not touch vendored code unless breaking

File: sktime/forecasting/timemoe.py (hunk 9 above)
Comment (geetu040) [reply sktime/forecasting/timemoe.py:630]: agreed, I think Option 1 is better

**Review by geetu040 — CHANGES_REQUESTED** [review]
Nice. One thing though, it's missing `validation_split` which I think is very important.

**Review by geetu040 — DISMISSED** [review]
LGTM!

CC: @fkiraly for review/merge.

**Review by geetu040 — APPROVED** [review]
@fkiraly we can merge this as it is before the release `1.2.0` otherwise we need some changes for backward compatibility.
